sdk/storage/AzureStorageBlob/Source/StorageBlobClient.swift

```diff
@@ -133,9 +133,12 @@ public class StorageBlobClient: PipelineClient, PagedCollectionDelegate {
             return try self.init(accountUrl: blobEndpoint, credential: sasCredential, withOptions: options)
         }
 
-    // MARK: Private Methods
+    public static func url(forHost host: String, container: String, blob: String) -> URL? {
+        let urlString = "\(host)/\(container)/\(blob)".addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed)
```

Comment:
The URLSessionTransport will perform the percent-encoding before dispatching the request, do we also need to do it here?

Reply:
Yes. The URL initializer will fail if the string is not percent encoded. I suppose I could convert the URL objects to strings and see what repercussions that would have.

Reply:
The logic is all plumbed through with URL including the Core Data models, which does require the string to be percent encoded.

Refactoring from URL to String introduces a bunch of extra work. If it would truly be valuable to make this change, I would recommend create a stand-alone issue to be addressed separately.

Reply:
I don't think it's worth changing from URL to string, not least because it reduces fidelity of the parameter's type. In my experience, encoding becomes a pretty rich source of bugs because it's very difficult to tell if you're dealing with an encoded URL or not, and you can easily accidentally wind up double-encoding things.

Based on what you mentioned (the initializer will fail if not encoded) - can we assume that any instance of URL *must* be encoded already? If that's the case, that helps resolve this complexity by a ton.

Reply:
Yes, that should be the case. Actually, by the same logic it might make sense to make the URL parameter in `HTTPRequest` a URL object instead of a string. I will explore that possibility. Then we are consistently using URLs all around.